"Unable to start gopls with multiple nested directories."

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

If the opened directory is 6.5840, the LSP not work , and there are no code suggestions. If the opened directory is 'src', everything works fine.

project： `git clone git://g.csail.mit.edu/6.5840-golabs-2024 6.5840` 

![image](https://github.com/zed-industries/zed/assets/16971804/e9c72c6a-1a82-4d11-9f3b-219b92598f6e)


my settings

```json
// Zed settings
//
// For information on how to configure Zed, see the Zed
// documentation: https://zed.dev/docs/configuring-zed
//
// To see all of Zed's default settings without changing your
// custom settings, run the `open default settings` command
// from the command palette or from `Zed` application menu.
{
  "features": {
    "copilot": false
  },
  "theme": "One Dark",
  "telemetry": {
    "metrics": false
  },
  "autosave": {
    "after_delay": {
      "milliseconds": 3000
    }
  },
  "vim_mode": true,
  "ui_font_size": 18,
  "ui_font_family": "JetBrainsMono Nerd Font",
  "buffer_font_size": 18,
  "buffer_font_family": "JetBrainsMono Nerd Font Propo",
  "format_on_save": "off",
  "inlay_hints": {
    "enabled": false,
    "show_type_hints": true,
    "show_parameter_hints": true,
    "show_other_hints": true
  },
  "lsp": {
    "gopls": {
      "binary": {
        "path": "/Users/yangxiaohui/go/bin/gopls"
      },
      "initialization_options": {
        "hints": {
          "assignVariableTypes": true,
          "compositeLiteralFields": true,
          "compositeLiteralTypes": true,
          "constantValues": true,
          "functionTypeParameters": true,
          "parameterNames": true,
          "rangeVariableTypes": true
        }
      },
      "ui.semanticTokens": true,
      "symbolStyle": "Dynamic"
    }
  }
}
```




### Environment

Zed: v0.129.2 (Zed)
OS: macOS 14.4.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

2024-04-05T22:16:47-07:00 [INFO] ========== starting zed ==========
2024-04-05T22:16:47-07:00 [INFO] Opening main db
2024-04-05T22:16:47-07:00 [INFO] set environment variables from shell:/opt/homebrew/bin/fish, path:/opt/homebrew/sbin:/Users/yangxiaohui/go/bin:/opt/homebrew/opt/go@1.20/bin:/opt/homebrew/bin/:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin
2024-04-05T22:16:47-07:00 [INFO] Opening main db
2024-04-05T22:16:47-07:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-04-05T22:16:47-07:00 [INFO] Opening main db
2024-04-05T22:16:47-07:00 [INFO] Opening main db
2024-04-05T22:16:47-07:00 [ERROR] crates/workspace/src/persistence/model.rs:268: No worktree for path: "/Users/yangxiaohui/.config/zed/settings.json"
2024-04-05T22:16:48-07:00 [INFO] set status on client 0: Authenticating
2024-04-05T22:16:48-07:00 [INFO] set status on client 65188: Connecting
2024-04-05T22:16:48-07:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2024-04-05T22:16:49-07:00 [INFO] add connection to peer
2024-04-05T22:16:49-07:00 [INFO] waiting for server hello
2024-04-05T22:16:49-07:00 [INFO] got server hello
2024-04-05T22:16:49-07:00 [INFO] set status to connected (connection id: ConnectionId { owner_id: 0, id: 0 }, peer id: PeerId { owner_id: 387, id: 1118833 })
2024-04-05T22:16:49-07:00 [INFO] set status on client 65188: Connected { peer_id: PeerId { owner_id: 387, id: 1118833 }, connection_id: ConnectionId { owner_id: 0, id: 0 } }
2024-04-05T22:16:56-07:00 [INFO] installing extension make latest version
2024-04-05T22:16:57-07:00 [INFO] rebuilt extension index in 3.762ms
2024-04-05T22:16:57-07:00 [INFO] extensions updated. loading 1, reloading 0, unloading 0
2024-04-05T22:17:02-07:00 [INFO] installing extension git-firefly latest version
2024-04-05T22:17:02-07:00 [INFO] rebuilt extension index in 4.736166ms
2024-04-05T22:17:02-07:00 [INFO] extensions updated. loading 1, reloading 0, unloading 0
2024-04-05T22:17:08-07:00 [INFO] starting language server "gopls", path: "/Users/yangxiaohui/mit6824/project", id: 1
2024-04-05T22:17:08-07:00 [INFO] found user-installed language server for Go. path: "/Users/yangxiaohui/go/bin/gopls", arguments: []
2024-04-05T22:17:08-07:00 [INFO] starting language server. binary path: "/Users/yangxiaohui/go/bin/gopls", working directory: "/Users/yangxiaohui/mit6824/project", args: []
2024-04-05T22:17:08-07:00 [INFO] Language server with id 1 sent unhandled notification client/unregisterCapability:
{
  "unregisterations": [
    {
      "id": "workspace/didChangeWatchedFiles-0",
      "method": "workspace/didChangeWatchedFiles"
    }
  ]
}

--- a/crates/project/src/project.rs
+++ b/crates/project/src/project.rs
@@ -4283,7 +4283,6 @@ impl Project {
         let Some(watchers) = self
             .language_server_watcher_registrations
             .get(&language_server_id)
-            .map(|registrations| registrations)
         else {
             return;
         };

--- a/crates/project/src/project.rs
+++ b/crates/project/src/project.rs
@@ -55,9 +55,9 @@ use language::{
 use log::error;
 use lsp::{
     DiagnosticSeverity, DiagnosticTag, DidChangeWatchedFilesRegistrationOptions,
-    DocumentHighlightKind, Edit, LanguageServer, LanguageServerBinary, LanguageServerId,
-    LspRequestFuture, MessageActionItem, OneOf, ServerCapabilities, ServerHealthStatus,
-    ServerStatus, TextEdit,
+    DocumentHighlightKind, Edit, FileSystemWatcher, LanguageServer, LanguageServerBinary,
+    LanguageServerId, LspRequestFuture, MessageActionItem, OneOf, ServerCapabilities,
+    ServerHealthStatus, ServerStatus, TextEdit,
 };
 use lsp_command::*;
 use node_runtime::NodeRuntime;
@@ -167,6 +167,8 @@ pub struct Project {
     last_formatting_failure: Option<String>,
     last_workspace_edits_by_language_server: HashMap<LanguageServerId, ProjectTransaction>,
     language_server_watched_paths: HashMap<LanguageServerId, HashMap<WorktreeId, GlobSet>>,
+    language_server_watcher_registrations:
+        HashMap<LanguageServerId, HashMap<String, Vec<FileSystemWatcher>>>,
     client: Arc<client::Client>,
     next_entry_id: Arc<AtomicUsize>,
     join_project_response_message_id: u32,
@@ -725,6 +727,7 @@ impl Project {
                 last_formatting_failure: None,
                 last_workspace_edits_by_language_server: Default::default(),
                 language_server_watched_paths: HashMap::default(),
+                language_server_watcher_registrations: HashMap::default(),
                 buffers_being_formatted: Default::default(),
                 buffers_needing_diff: Default::default(),
                 git_diff_debouncer: DebouncedDelay::new(),
@@ -875,6 +878,7 @@ impl Project {
                 last_formatting_failure: None,
                 last_workspace_edits_by_language_server: Default::default(),
                 language_server_watched_paths: HashMap::default(),
+                language_server_watcher_registrations: HashMap::default(),
                 opened_buffers: Default::default(),
                 buffers_being_formatted: Default::default(),
                 buffers_needing_diff: Default::default(),
@@ -3471,10 +3475,31 @@ impl Project {
                                     let options = serde_json::from_value(options)?;
                                     this.update(&mut cx, |this, cx| {
                                         this.on_lsp_did_change_watched_files(
-                                            server_id, options, cx,
+                                            server_id, &reg.id, options, cx,
                                         );
                                     })?;
                                 }
+                            }
+                        }
+                        Ok(())
+                    }
+                }
+            })
+            .detach();
+
+        language_server
+            .on_request::<lsp::request::UnregisterCapability, _, _>({
+                let this = this.clone();
+                move |params, mut cx| {
+                    let this = this.clone();
+                    async move {
+                        for unreg in params.unregisterations.iter() {
+                            if unreg.method == "workspace/didChangeWatchedFiles" {
+                                this.update(&mut cx, |this, cx| {
+                                    this.on_lsp_unregister_did_change_watched_files(
+                                        server_id, &unreg.id, cx,
+                                    );
+                                })?;
                             }
                         }
                         Ok(())
@@ -4208,16 +4233,68 @@ impl Project {
     fn on_lsp_did_change_watched_files(
         &mut self,
         language_server_id: LanguageServerId,
+        registration_id: &str,
         params: DidChangeWatchedFilesRegistrationOptions,
         cx: &mut ModelContext<Self>,
     ) {
+        let registrations = self
+            .language_server_watcher_registrations
+            .entry(language_server_id)
+            .or_default();
+
+        registrations.insert(registration_id.to_string(), params.watchers);
+
+        self.rebuild_watched_paths(language_server_id, cx);
+    }
+
+    fn on_lsp_unregister_did_change_watched_files(
+        &mut self,
+        language_server_id: LanguageServerId,
+        registration_id: &str,
+        cx: &mut ModelContext<Self>,
+    ) {
+        let registrations = self
+            .language_server_watcher_registrations
+            .entry(language_server_id)
+            .or_default();
+
+        if registrations.remove(registration_id).is_some() {
+            log::info!(
+                "language server {}: unregistered workspace/DidChangeWatchedFiles capability with id {}",
+                language_server_id,
+                registration_id
+            );
+        } else {
+            log::warn!(
+                "language server {}: failed to unregister workspace/DidChangeWatchedFiles capability with id {}. not registered.",
+                language_server_id,
+                registration_id
+            );
+        }
+
+        self.rebuild_watched_paths(language_server_id, cx);
+    }
+
+    fn rebuild_watched_paths(
+        &mut self,
+        language_server_id: LanguageServerId,
+        cx: &mut ModelContext<Self>,
+    ) {
+        let Some(watchers) = self
+            .language_server_watcher_registrations
+            .get(&language_server_id)
+            .map(|registrations| registrations)
+        else {
+            return;
+        };
+
         let watched_paths = self
             .language_server_watched_paths
             .entry(language_server_id)
             .or_default();
 
         let mut builders = HashMap::default();
-        for watcher in params.watchers {
+        for watcher in watchers.values().flatten() {
             for worktree in &self.worktrees {
                 if let Some(worktree) = worktree.upgrade() {
                     let glob_is_inside_worktree = worktree.update(cx, |tree, _| {

